### ISSUE TITLE
Reusable onChange function

### ISSUE THREAD

**Issue by susuwatari2** [JedWatson/react-select#3206]
### Thanks for using react-select!

Before creating an issue...

# Are you asking a question?
Please don't file GitHub issues to ask questions. Use Stack Overflow with a [#react-select](http://stackoverflow.com/questions/tagged/react-select) tag

# Are you reporting a bug or runtime error?
Please include a test case that demonstrates the issue you're reporting!
This is very helpful to maintainers in order to help us see the issue you're seeing.

To report bugs against react-select v2 please fork the following code-sandbox:
https://codesandbox.io/s/k5rvn9z3lv

To report bugs against react-select v1 please fork the following code-sandbox:
https://codesandbox.io/s/n34mq0743j

You may also find the [online Babel tool](https://babeljs.io/repl/) quite helpful if you wish to use ES6/ES7 syntax not yet supported by the browser you are using.

# Are you making a feature request?
Provide as much information as possible about your requested feature. Here are a few questions you may consider answering:

*   What's your use case? (Tell us about your application and what problem you're trying to solve.)
*   What interface do you have in mind? (What new properties or methods do you think might be helpful?)
*   Can you point to similar functionality with any existing libraries or components? (Working demos can be helpful.)

**Comment by cutterbl** [comment]
According to the docs, the `onChange` method gets two arguments, `option`, and `{action}`. It does not receive the event.

`<Select {...otherProps} onChange={(option, {action}) => doSomethingWith(option, action)} />`

If you're wrapping the Select, then your component `this` scope is relevant to maintaining your `target`.

**Comment by susuwatari2** [comment]
With `<Select>` I cannot get my target. I created an [example](https://codesandbox.io/s/qz8o7w7929) and get `onChange` working. Is it a better way to reuse the `onChange` method or update the state with the `defaultValue`?

Thank you.

**Comment by Rall3n** [comment]
@susuwatari2 

To update your state for each Select with the same function, you could do it like this:
```jsx
class MyComponent extends Component {
    {...}

    onChange = (opt, name) => {
        this.setState({ [name]: opt.value });
    }
    
    render() {
        return <Select onChange={(opt) => this.onChange(opt, 'color')} {...} />
    }
}
```

**Comment by susuwatari2** [comment]
Thank you @Rall3n !! It works.

How come my `defaultValue `is not work? Does it matter with the react life cycle?

**Comment by Rall3n** [comment]
As @gwyneplaine mentioned in the issue:
The `defaultValue` and `value` props both expect objects in the shape of
```ts
{ value: string | number, label: string }
```

So you have to give them the whole option objects. For this, you can use [Array.prototype.find](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/find), which is supported by almost all browsers except IE, which you might want to polyfill, if you have to.

```jsx
<Select defaultValue={options.find(elem => elem.value === this.state.value)} {...} />
```

But I would recommend using `value` instead of `defaultValue`, as the latter only sets the value for first render / mount, and `value` will keep the state and the selected value always synchronized.

**Comment by waynebloss** [comment]
Every other type of input in React pretty much passes `target` as a sub-property of the first `onChange` argument.

Except react-select. 🤕

**Comment by kivervinicius** [comment]
@waynebloss 

`onChange={option => {
                let val = "";
                if (option) {
                  val = option.value;
                }
                this.handleChange.call(this, {
                  target: {
                    name: "produtoId",
                    value: val
                  }
                });
              }}`

**Comment by waynebloss** [comment]
@kivervinicius Yes, thanks. Been there, done that. That code is so trivial (well outside of your unusual `.call` semantics) that perhaps it should have just been built into the component eh?

Surely that's better than hauling around 13 lines of boilerplate code every time you want to use a select box, lmao!

**Comment by samuely4ng** [comment]
thank you so much @Rall3n, I've been looking for this for quite a while!

**Comment by nirajrajgor** [comment]
In the version 2.0.11 and above to update your state for each Select with the same function, you could do it like this:
```
class MyComponent extends Component{
 this.state={
  sortBy: null
 }
 handleSelectChange = (selectedOption, { name }) => {
   this.setState({ [name]: selectedOption.value });
 }
 render() {
      return <Select name="sortBy" onChange={this.handleSelectChange} {...} />
 }
}
```
